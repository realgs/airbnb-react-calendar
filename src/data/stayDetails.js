export const stayDetails = {
  user: 'Matthew.Jings',
  price: 19.9,
  rating : 3.5,
  reviews : 123,
  bonus : 10,
  cleaningFee : 10,
  stayLength : 0,
  minStay: 2,
};