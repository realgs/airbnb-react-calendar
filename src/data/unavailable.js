export const unavailableDates = [
  "2018-11-01",
  "2018-11-02",
  "2018-11-14",
  "2018-11-16",
  "2018-11-20",
  "2018-11-25",
  "2018-11-26",
  "2018-12-01",
  "2018-12-02",
  "2018-12-03",
  "2018-12-15",
  "2018-12-16",
  "2018-12-17",
  "2019-01-05",
  "2019-01-06",
  "2019-01-07",
  "2019-01-08",
  "2019-01-15",
  "2019-01-16",
  "2019-01-17",
  "2019-01-20",
  "2019-01-21",
  "2019-01-22",
  "2019-01-26"
];