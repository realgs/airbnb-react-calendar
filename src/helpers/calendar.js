export const ONE_DAY = 1000 * 60 * 60 * 24;
export const round2Decimals = (num) => {
  return Math.round(num * 100 + Number.EPSILON) / 100;
};

export const THIS_YEAR = +(new Date().getFullYear());
export const THIS_MONTH = +(new Date().getMonth());

export const WEEK_DAYS = [
  "Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"
];

export const MONTHS = ["January", "February", "March", "April", "May", "June",
  "July", "August", "September", "October", "November", "December"
];

export const zeroPad = (value, length) => {
  return `${value}`.padStart(length, '0');
}
// 1 - Sunday, ..., 7 - Saturday
export const getMonthFirstDay = (month = THIS_MONTH, year = THIS_YEAR) => {
  return +(new Date(`${year}-${zeroPad(month + 1, 2)}-01`).getDay());
}

export const NO_DAYS_MONTH = getMonthDays();

export function getMonthDays(month = THIS_MONTH, year = THIS_YEAR) {
  return 32 - new Date(year, month, 32).getDate();
}

export const thisMonthDates = (month = THIS_MONTH, year = THIS_YEAR) => {
  const daysToIndent = getMonthFirstDay(month, year);
  let indents = [...new Array(daysToIndent)].map((n, index) => {
    const day = index + 1;
    return ['indent', zeroPad(day, 2)];
  });
  const daysOfMonth = [...new Array(getMonthDays(month, year))].map((n, index) => {
    const day = index + 1;
    return [year, zeroPad(month + 1, 2), zeroPad(day, 2)];
  });
  const daysToPad = 7 - (indents.length + daysOfMonth.length) % 7;
  const pads = [...new Array(daysToPad)].map((n, index) => {
    const day = index + 1;
    return ['pad', zeroPad(day, 2)];
  });
  const monthDates = [...indents, ...daysOfMonth, ...pads];
  let monthDatesDivided = [];
  for (var i = 0; i < monthDates.length; i += 7)
    monthDatesDivided.push(monthDates.slice(i, i + 7));
  return monthDatesDivided;
};

// (bool) Checks if a value is a date
export const isDate = date => {
  const isDate = Object.prototype.toString.call(date) === '[object Date]';
  const isValidDate = date && !Number.isNaN(date.valueOf());

  return isDate && isValidDate;
}

// (bool) Checks if two date values are of the same month and year
export const isSameMonth = (date, basedate = new Date()) => {

  if (!(isDate(date) && isDate(basedate))) return false;

  const basedateMonth = +(basedate.getMonth());
  const basedateYear = basedate.getFullYear();

  const dateMonth = +(date.getMonth());
  const dateYear = date.getFullYear();

  return (+basedateMonth === +dateMonth) && (+basedateYear === +dateYear);

}

// (bool) Checks if two date values are the same day
export const isSameDay = (date, basedate = new Date()) => {

  if (!(isDate(date) && isDate(basedate))) return false;

  const basedateDate = basedate.getDate();
  const basedateMonth = +(basedate.getMonth());
  const basedateYear = basedate.getFullYear();

  const dateDate = date.getDate();
  const dateMonth = +(date.getMonth());
  const dateYear = date.getFullYear();

  return (+basedateDate === +dateDate) && (+basedateMonth === +dateMonth) && (+basedateYear === +dateYear);

}

export const getDateISO = (date = new Date) => {

  if (!isDate(date)) return null;

  return [
    date.getFullYear(),
    zeroPad(+date.getMonth() + 1, 2),
    zeroPad(+date.getDate(), 2)
  ].join('-');

}

export const getPreviousMonth = (prevMonth, prevYear) => {
  const newMonth = (prevMonth > 0) ? prevMonth - 1 : 11;
  const newYear = (prevMonth > 0) ? prevYear : prevYear - 1;

  return { month: newMonth, year: newYear };
}


export const getNextMonth = (month, year) => {
  const nextMonth = (month < 11) ? month + 1 : 0;
  const nextMonthYear = (month < 11) ? year : year + 1;

  return { month: nextMonth, year: nextMonthYear };
}
// Calendar builder for a month in the specified year
// Returns an array of the calendar dates.
// Each calendar date is represented as an array => [YYYY, MM, DD]

// export default (month = THIS_MONTH, year = THIS_YEAR) => {

//   // Get number of days in the month and the month's first day

//   const monthDays = getMonthDays(month, year);
//   const monthFirstDay = getMonthFirstDay(month, year);

//   // Get number of days to be displayed from previous and next months
//   // These ensure a total of 42 days (6 weeks) displayed on the calendar

//   // Get the previous and next months and years

//   const { month: prevMonth, year: prevMonthYear } = getPreviousMonth(month, year);
//   const { month: nextMonth, year: nextMonthYear } = getNextMonth(month, year);


//   // Builds dates to be displayed from current month

//   const thisMonthDates = [...new Array(monthDays)].map((n, index) => {
//     const day = index + 1;
//     return [year, zeroPad(month, 2), zeroPad(day, 2)];
//   });

//   // Combines all dates from previous, current and next months
//   return [...prevMonthDates, ...thisMonthDates];

// }
