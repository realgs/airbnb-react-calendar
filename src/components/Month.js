import React from 'react';
import { getMonthFirstDay, thisMonthDates, getDateISO } from '../helpers/calendar';
import { WEEK_DAYS } from '../helpers/calendar';

export default class Month extends React.Component {
  constructor(props) {
    super(props);
    this.renderDay = this.renderDay.bind(this);
    this.renderWeeks = this.renderWeeks.bind(this);
    this.canSet = this.canSet.bind(this);
    this.state = {
      from: null,
      to: null,
      lastModified: null,
      reset: this.props.reset,
      current: this.props.current,
      naDays: [],
    };
  }
  componentDidMount() {
  }
  componentDidUpdate(prevProps, prevState) {
    if (this.state.from) {
      document.getElementById("checkin").innerHTML = this.state.from;
      if (this.state.to) {
        document.getElementById("checkout").innerHTML = this.state.to;
        let fromDate = new Date(this.state.from);
        let toDate = new Date(this.state.to);
        if (fromDate > toDate) {
          //console.log(`-----Switched days from: ${fromDate.toLocaleDateString()} --- ${toDate.toLocaleDateString()}`);
          //console.log(`To: ${toDate.toLocaleDateString()} --- ${fromDate.toLocaleDateString()}`);
          this.setState({
            from: prevState.to ? prevState.to : this.state.to,
            to: prevState.from ? prevState.from : this.state.from
          });
        }
      }
    }
  }
  handleSetDay(day) {

    if (this.state.reset) {
      this.setState({
        from: day,
        lastModified: day,
        to: null,
        reset: false,
      });
      this.props.toggleReset();
    } else {
      if (this.canSet(day)) {
        this.setState({ to: day, lastModified: day, reset: true });
        this.props.toggleReset();
      } else {
        console.log('Sorry these days are unavailable.')
      }
    }

  }
  canSet(day) {
    const minMax = [this.state.lastModified, day].sort();
    const naDays = this.props.unavailable.filter((d) => {
      return  d > minMax[0] && d < minMax[1];
    });
    this.setState({ naDays });
    return !Array.isArray(naDays) || !naDays.length;
  }

  renderDay(day, index) {
    const thisDay = new Date(day);
    const isoDate = getDateISO(thisDay);
    const dayNA = this.props.unavailable.indexOf(isoDate) > -1;
    const daySelected = this.state.from == isoDate || this.state.to == isoDate;
    const dayBetween = this.state.from < isoDate && this.state.to > isoDate;
    const dayConflict = this.state.naDays.indexOf(isoDate) > -1;
    if (day[0] === 'indent' || day[0] === 'pad') {
      return (
        <td key={index} id={index} className="day day__blank"></td>
      );
    } else {
      return (
        <td key={isoDate} id={isoDate} className="day">
          <button
          id={isoDate}
          className={`day__button${daySelected || dayBetween ? ' day__button__selected' : ''}${dayNA ? ' day__button__na' : ''}${dayConflict ? ' day__button__conflict' : ''}`}
          disabled={daySelected || dayNA} onClick={(e) => { this.handleSetDay(e.target.id) }}
          >
            {thisDay.getDate()}
          </button>
        </td>
      );
    }
  }
  renderWeeks(week, weekIndex) {
    return (
      <tr key={weekIndex} id={weekIndex}>
        {week.map(this.renderDay)}
      </tr>
    );
  }
  renderWeekDays(name, index) {
    return (
      <th key={index} className="month__weekday">
        {name}
      </th>
    );
  }

  render() {
    return (
      <div className="month">
        <table className="month__table">
          <thead>
            <tr>
              {WEEK_DAYS.map(this.renderWeekDays)}
            </tr>
          </thead>
          <tbody>
            {thisMonthDates(this.props.current.month, this.props.current.year).map(this.renderWeeks)}
          </tbody>
        </table>
      </div>
    );
  }
}