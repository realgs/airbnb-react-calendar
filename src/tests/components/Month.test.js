import React from 'react';
import { shallow } from 'enzyme';
import Month from '../../components/Month';
import { unavailableDates } from '../../data/unavailable';
import { THIS_YEAR, THIS_MONTH } from '../../helpers/calendar';


test('Should match Month snapshot.', () => {
  const current = { month: THIS_MONTH, year: THIS_YEAR };
  const handleSetDay = jest.fn();
  const handleSetCandidate = jest.fn();
  const handleDeleteCandidate = jest.fn();
  const wrapper = shallow(<Month
    current={current}
    unavailable={unavailableDates}
    handleSetDay={handleSetDay}
    handleSetCandidate={handleSetCandidate}
    handleDeleteCandidate={handleDeleteCandidate}
  />);
  expect(wrapper).toMatchSnapshot();
});