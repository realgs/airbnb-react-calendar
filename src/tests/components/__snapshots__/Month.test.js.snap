// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match Month snapshot. 1`] = `
<div
  className="month"
>
  <table
    className="month__table"
  >
    <thead>
      <tr>
        <th
          className="month__weekday"
          key="0"
        >
          Su
        </th>
        <th
          className="month__weekday"
          key="1"
        >
          Mo
        </th>
        <th
          className="month__weekday"
          key="2"
        >
          Tu
        </th>
        <th
          className="month__weekday"
          key="3"
        >
          We
        </th>
        <th
          className="month__weekday"
          key="4"
        >
          Th
        </th>
        <th
          className="month__weekday"
          key="5"
        >
          Fr
        </th>
        <th
          className="month__weekday"
          key="6"
        >
          Sa
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        id={0}
        key="0"
      >
        <td
          className="day day__blank"
          id={0}
          key="0"
        />
        <td
          className="day day__blank"
          id={1}
          key="1"
        />
        <td
          className="day"
          id="2019-01-01"
          key="2019-01-01"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-01"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            1
          </button>
        </td>
        <td
          className="day"
          id="2019-01-02"
          key="2019-01-02"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-02"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            2
          </button>
        </td>
        <td
          className="day"
          id="2019-01-03"
          key="2019-01-03"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-03"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            3
          </button>
        </td>
        <td
          className="day"
          id="2019-01-04"
          key="2019-01-04"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-04"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            4
          </button>
        </td>
        <td
          className="day"
          id="2019-01-05"
          key="2019-01-05"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-05"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            5
          </button>
        </td>
      </tr>
      <tr
        id={1}
        key="1"
      >
        <td
          className="day"
          id="2019-01-06"
          key="2019-01-06"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-06"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            6
          </button>
        </td>
        <td
          className="day"
          id="2019-01-07"
          key="2019-01-07"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-07"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            7
          </button>
        </td>
        <td
          className="day"
          id="2019-01-08"
          key="2019-01-08"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-08"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            8
          </button>
        </td>
        <td
          className="day"
          id="2019-01-09"
          key="2019-01-09"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-09"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            9
          </button>
        </td>
        <td
          className="day"
          id="2019-01-10"
          key="2019-01-10"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-10"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            10
          </button>
        </td>
        <td
          className="day"
          id="2019-01-11"
          key="2019-01-11"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-11"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            11
          </button>
        </td>
        <td
          className="day"
          id="2019-01-12"
          key="2019-01-12"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-12"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            12
          </button>
        </td>
      </tr>
      <tr
        id={2}
        key="2"
      >
        <td
          className="day"
          id="2019-01-13"
          key="2019-01-13"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-13"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            13
          </button>
        </td>
        <td
          className="day"
          id="2019-01-14"
          key="2019-01-14"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-14"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            14
          </button>
        </td>
        <td
          className="day"
          id="2019-01-15"
          key="2019-01-15"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-15"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            15
          </button>
        </td>
        <td
          className="day"
          id="2019-01-16"
          key="2019-01-16"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-16"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            16
          </button>
        </td>
        <td
          className="day"
          id="2019-01-17"
          key="2019-01-17"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-17"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            17
          </button>
        </td>
        <td
          className="day"
          id="2019-01-18"
          key="2019-01-18"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-18"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            18
          </button>
        </td>
        <td
          className="day"
          id="2019-01-19"
          key="2019-01-19"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-19"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            19
          </button>
        </td>
      </tr>
      <tr
        id={3}
        key="3"
      >
        <td
          className="day"
          id="2019-01-20"
          key="2019-01-20"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-20"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            20
          </button>
        </td>
        <td
          className="day"
          id="2019-01-21"
          key="2019-01-21"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-21"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            21
          </button>
        </td>
        <td
          className="day"
          id="2019-01-22"
          key="2019-01-22"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-22"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            22
          </button>
        </td>
        <td
          className="day"
          id="2019-01-23"
          key="2019-01-23"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-23"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            23
          </button>
        </td>
        <td
          className="day"
          id="2019-01-24"
          key="2019-01-24"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-24"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            24
          </button>
        </td>
        <td
          className="day"
          id="2019-01-25"
          key="2019-01-25"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-25"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            25
          </button>
        </td>
        <td
          className="day"
          id="2019-01-26"
          key="2019-01-26"
        >
          <button
            className="day__button day__button__na"
            disabled={true}
            id="2019-01-26"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            26
          </button>
        </td>
      </tr>
      <tr
        id={4}
        key="4"
      >
        <td
          className="day"
          id="2019-01-27"
          key="2019-01-27"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-27"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            27
          </button>
        </td>
        <td
          className="day"
          id="2019-01-28"
          key="2019-01-28"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-28"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            28
          </button>
        </td>
        <td
          className="day"
          id="2019-01-29"
          key="2019-01-29"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-29"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            29
          </button>
        </td>
        <td
          className="day"
          id="2019-01-30"
          key="2019-01-30"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-30"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            30
          </button>
        </td>
        <td
          className="day"
          id="2019-01-31"
          key="2019-01-31"
        >
          <button
            className="day__button"
            disabled={false}
            id="2019-01-31"
            onClick={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
          >
            31
          </button>
        </td>
        <td
          className="day day__blank"
          id={5}
          key="5"
        />
        <td
          className="day day__blank"
          id={6}
          key="6"
        />
      </tr>
    </tbody>
  </table>
</div>
`;
